FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -125,102 +127,148 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
+                NSDictionary *dict = [self getDictionaryFromArrayObject:arr withKey:key];
                 [props addEntriesFromDictionary:dict];
-            }else{
+            } else {
+                // if this is not a dictionary or array, then treat it like simple values, if data falls outside of these, we don't try to infer anything and just send as is to the server
                 suffix = [self getSuffixStringFromSimpleObject:obj];
-                [props setValue:[properties valueForKey:key] forKey:[key stringByAppendingString:suffix]];
+                [self appendToDictionary:props withKey:[key stringByAppendingString:suffix] andValue:properties[key]];
             }
         }else{
             #ifdef DEBUG
                 NSAssert(FALSE, @"key `%@` is not an object can't be serialized for FS custom event.", key);
             #else
                 NSLog(@"key `%@` is not an object can't be serialized for FS custom event.", key);
-                // if prod, then don't send it
+                // if prod, then ignore this pops all together
             #endif
         }
     }
+
     return props;
 }
 
-- (NSString *) getSuffixStringFromSimpleObject: (NSObject *) obj{
-    NSString * suffix;
-    // defualt to string
-    if([obj isKindOfClass:[NSNumber class]]){
+- (NSString *) getSuffixStringFromSimpleObject: (NSObject *) obj {
+    // default to no suffix;
+    NSString * suffix = @"";
+    if ([obj isKindOfClass:[NSNumber class]]) {
         // defaut to real
         suffix = @"_real";
         NSNumber *n = (NSNumber *) obj;
         const char *type = n.objCType;
-        if([@"i" isEqualToString:@(type)]) suffix = @"_int";
-        // bool type number doesn't get encoded into 'B' but check anyway
-        else if ([@"B" isEqualToString:@(type)])suffix = @"_bool";
-    }else if ([obj isKindOfClass:[NSDate class]]){
+        if([@"i" isEqualToString:@(type)]) {
+            suffix = @"_int";
+        } else if ([@"B" isEqualToString:@(type)]) {
+            // bool-type gets encoded as number number, and doesn't get encoded into 'B', but check anyway
+            suffix = @"_bool";
+        }
+    } else if ([obj isKindOfClass:[NSDate class]]) {
         suffix = @"_date";
-    }else if([obj isKindOfClass:[NSString class]]){
+    } else if([obj isKindOfClass:[NSString class]]) {
         suffix = @"_str";
     }
+
     return suffix;
 }
 
-- (NSDictionary *) getObjectFromArrayObject: (NSArray *) arr withKey:(NSString *) key{
-    // FS does not accept array of dicts, only premitive arrays allowed
-    // So if it's an array of dicts, then we need to convert the root array to a dict so it becomes nested dicts
+- (NSDictionary *) getDictionaryFromArrayObject:(NSArray *) arr withKey:(NSString *) key {
+    // FS does not accept array of dicts, only "premitive" arrays allowed
+    // So if it's an array of dicts or nested arrays, then we need to convert the root array to a dict so it becomes nested dicts
```

Comment:
The `if` in the comment above implies to me that the code needs to do a check, but it doesn't, it just arbitrarily converts this array into a dictionary, so I feel that the comment is misleading. Perhaps rewording to something like:
`Convert the root array into a dict since we don't support arrays with nested dicts or nested arrays.`